docs/incubator-proposal.md

```diff
@@ -0,0 +1,188 @@
+came about in order to automate the retrieval of certificates from an [ACME](https://github.com/ietf-wg-acme/acme/)
+compliant server. This has served the community well, especially with the rise
+in popularity of ACME due to [letsencrypt](https://letsencrypt.org). These
+projects also attempt to renew the certificate as it reaches its expiry time.
+
+Whilst each project has invented its own schema/methodology for obtaining
+these certificates, the end goal has largely been the same. To provide a signed
+certificate and private key in a Secret resource to then be consumed, either by
+an Ingress controller or some other workload in-cluster, and then keep that
+certificate up to date.
+
+Furthermore, it is not currently possible with any of these projects to support
+a mode of operation where the private key is kept secret at all times (i.e. the
+key never leaves the application container). It'd also be beneficial for users
+to be able to utilise alternative issuer backends, other than ACME. This could
+include [Hashicorp Vault](https://vaultproject.io), a plain CA (backed by
+Kubernetes secret resources), or any other cloud based service.
+
+It is therefore proposed that a new project, `cert-manager`, is created that
+aims to generalise the issuance and renewal of certificates from an arbitrary
+certificate source. This project should be designed in a similar manner to
+other Kubernetes controllers (eg. kube-controller-manager, service-catalog et
+al.). Whether certificates are signed by a publicly trusted authority is not a
+concern of cert-manager.
+
+cert-manager would also like to support this operating mode where the private
+key is never exposed outside of the application requesting a certificate.
+
+## Definitions
+
+* `Issuer` - an Issuer is a generic backend that issues certificates. cert-manager
+will contain logic on how to process each issuer, and multiple Issuer resources
+utilising the same issuer implementation may exist (e.g. to allow issuing
+certificates from both letsencrypt production & staging). An example manifest for
+an Issuer resource:
+
+```yaml
+kind: Issuer
+metadata:
+  name: letsencrypt-staging
+spec:
+  acme:
+    # The ACME server URL
+    server: https://acme-staging.api.letsencrypt.org/directory
+    # Email address used for ACME registration
+    email: user@example.com
+    # Name of a secret used to store the ACME account private key
+    privateKey: letsncrypt-staging
+    # Enable the HTTP-01 challenge mechanism
+    http-01: {}
+    # Enable the DNS-01 challenge mechanism and provide one DNS provider
+    dns-01:
+      # Here we define a list of DNS-01 providers that can solve DNS challenges
+      providers:
+      # We define a provider named 'clouddns', with configuration for the
+      # clouddns challenge provider.
+      - name: clouddns-prod
+        clouddns:
+          # A secretKeyRef to the google cloud json service account
+          serviceAccount:
+            name: clouddns-service-account
```

Comment:
this approach makes me nervous since I imagine this account is likely a cluster to every single web presence escalation for most DNS providers. But if you want to automate the DNS challenge you need something with write access to DNS.